python/dgl/batched_heterograph.py

```diff
@@ -0,0 +1,410 @@
+    ...                       edge_attrs={('user', 'plays', 'game'): 'h1'})
+
+    Below one can see that the nodes are re-indexed. The edges are re-indexed in
+    the same way.
+
+    >>> bg.nodes('user')
+    tensor([0, 1, 2, 3, 4, 5])
+    >>> bg.nodes['user'].data['h1']
+    tensor([[0.],
+            [1.],
+            [2.],
+            [0.],
+            [1.],
+            [2.]])
+
+    **Property:**
+
+    We can still get a brief summary of the graphs that constitute the batched graph.
+
+    >>> bg.batch_size
+    2
+    >>> bg.batch_num_nodes('user')
+    [3, 3]
+    >>> bg.batch_num_edges(('user', 'plays', 'game'))
+    [2, 2]
+
+    **Update Attributes:**
+
+    Updating the attributes of the batched graph has no effect on the original graphs.
+
+    >>> bg.nodes['game'].data['h1'] = th.tensor([[1.], [1.]])
+    >>> g2.nodes['game'].data['h1']
+    tensor([[0.]])
+
+    Instead, we can decompose the batched graph back into a list of graphs and use them
+    to replace the original graphs.
+
+    >>> g3, g4 = dgl.unbatch_hetero(bg) # returns a list of DGLHeteroGraph objects
+    >>> g4.nodes['game'].data['h1']
+    tensor([[1.]])
+    """
+    def __init__(self, graph_list, node_attrs, edge_attrs):
+        # Sanity check. Make sure all graphs have the same node/edge types, in the same order.
+        ref_graph = graph_list[0]
+        ref_canonical_etypes = ref_graph.canonical_etypes
+        ref_ntypes = ref_graph.ntypes
+        ref_etypes = ref_graph.etypes
+        for i in range(1, len(graph_list)):
+            g_i = graph_list[i]
+            g_i_canonical_etypes = g_i.canonical_etypes
+            g_i_ntypes = g_i.ntypes
+            assert len(g_i_canonical_etypes) == len(ref_canonical_etypes), \
+                'Got different numbers of canonical etypes for graphs in the same batch.'
+            assert len(g_i_ntypes) == len(ref_ntypes), \
+                'Got different number of node types for graphs in the same batch.'
+            for tid, ety in enumerate(ref_canonical_etypes):
+                assert ety == g_i_canonical_etypes[tid], \
+                    'Got different canonical etype with id {:d} ' \
+                    'for graphs in the same batch.'.format(tid)
+            for tid, nty in enumerate(ref_ntypes):
+                assert nty == g_i_ntypes[tid], \
+                    'Got different node type with id {:d} ' \
+                    'for graphs in the same batch.'.format(tid)
+
+        # Merge node and edge frames
+        def _get_num_item_and_attr_types(g, mode, mode_type):
+            if mode == 'node':
+                num_items = g.number_of_nodes(mode_type)
+                attr_types = set(g.node_attr_schemes(mode_type).keys())
+            elif mode == 'edge':
+                num_items = g.number_of_edges(mode_type)
+                attr_types = set(g.edge_attr_schemes(mode_type).keys())
+            return num_items, attr_types
+
+        def _init_attrs(types, attrs, mode):
```

Comment:
Can we require all the input graphs to have the same attribute schema? Is there any scenarios where batching involves complex schema? I remembered there were some issues with graphs with no edges. But could we also require these graphs to have the same schema? I feel this will greatly simplify the logic, making it potentially easier to maintain (and to refactor in the future).

Reply:
Fine. Then do you think whether we need sanity check for node/edge types?

Reply:
We should. Is that easy?

Reply:
Done.